sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTimeWindows.scala

```diff
@@ -103,9 +115,10 @@ object TimeWindowing extends Rule[LogicalPlan] {
 
         def getWindow(i: Int, dataType: DataType): Expression = {
           val timestamp = PreciseTimestampConversion(window.timeColumn, dataType, LongType)
-          val lastStart = timestamp - (timestamp - window.startTime
-            + window.slideDuration) % window.slideDuration
-          val windowStart = lastStart - i * window.slideDuration
+          val lastStart = timestamp - (timestamp - window.startTime) % window.slideDuration
```

Comment:
This seems a little complicated, and need not involve CaseWhen. You just always want to subtract a nonnegative amount here.

```
val remainder = (timestamp - window.startTime) % window.slideDuration
val lastStart = timestamp - (if (remainder < 0) remainder + window.slideDuration else remainder)
```

Reply:
> and need not involve CaseWhen.
This is actually incorrect. Although `window.startTime` and `window.slideDuration` are `Long`, `timestamp` is `Expression`, and `remainder < 0` returns a `Predicate` not `boolean` so the complier complains. 

> You just always want to subtract a nonnegative amount here.
Sure I could do that. I'll update a version with `CaseWhen` with your logic.